### PR TITLE
Add canned tensor_forest inference

### PR DESCRIPTION
**Pull request by yupbank**
According to https://github.com/tensorflow/community/blob/master/rfcs/20180626-tensor-forest.md

This is a first step. And the whole process is tracked by https://github.com/tensorflow/tensorflow/issues/21830

We are adding an canned TensorForestClassifier and TensorForestRegressor into core.

cc @nataliaponomareva

### CODE REVIEW

File: tensorflow_estimator/python/estimator/canned/tensor_forest.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+from tensorflow.python.estimator import estimator
+from tensorflow.python.estimator.canned import head as head_lib
+from tensorflow.python.estimator.canned import prediction_keys
+from tensorflow.python.ops.losses import losses
+from tensorflow.python.summary import summary
+from tensorflow.python.feature_column import feature_column as feature_column_lib
+from tensorflow.python.ops import gen_tensor_forest_ops
+from tensorflow.python.ops import tensor_forest_ops, math_ops, array_ops
+
+# from tensorflow.python.util.tf_export import estimator_export
```
Comment (tanzhenyu) [inline tensorflow_estimator/python/estimator/canned/tensor_forest.py]: Remove.

File: tensorflow_estimator/python/estimator/canned/tensor_forest.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+      0., expected_squares - average_values * average_values)
+
+  estimator_spec.predictions[VARIANCE_LOGITS_KEY] = logits_variance
+  estimator_spec.predictions[
+      VARIANCE_AFTER_ACTIVATION_KEY] = var_after_activation
+
+  return estimator_spec
+
+
+# @estimator_export('estimator.TensorForestClassifier')
```
Comment (tanzhenyu) [inline tensorflow_estimator/python/estimator/canned/tensor_forest.py]: @karmel Should this have experimental tag?

File: tensorflow_estimator/python/estimator/canned/tensor_forest.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    else:
+      assert head.logits_dimension == n_classes, 'head logits_dimension: %s ' \
+          % head.logits_dimension + \
+          'should match n_classes: %s' % n_classes
+
+    assert all(isinstance(fc, feature_column_lib.DenseColumn)
+               for fc in feature_columns), \
+        'Only dense columns are supported for now'
+
+    assert all(len(fc.shape) == 1
```
Comment (tanzhenyu) [inline tensorflow_estimator/python/estimator/canned/tensor_forest.py]: Prefer to be explicit using a for loop, let the user know which feature column they provide is not DenseColumn.

File: tensorflow_estimator/python/estimator/canned/tensor_forest.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+          n_classes=n_classes,
+          weight_column=None,
+          label_vocabulary=label_vocabulary,
+          loss_reduction=losses.Reduction.SUM_OVER_BATCH_SIZE)
+    else:
+      assert head.logits_dimension == n_classes, 'head logits_dimension: %s ' \
+          % head.logits_dimension + \
+          'should match n_classes: %s' % n_classes
+
+    assert all(isinstance(fc, feature_column_lib.DenseColumn)
```
Comment (tanzhenyu) [inline tensorflow_estimator/python/estimator/canned/tensor_forest.py]: I think sometime ago we added a check to see if it's DenseColumn from either feature_column or feature_column_V2.

File: tensorflow_estimator/python/estimator/canned/tensor_forest.py (hunk 2 above)
Comment (karmel) [reply tensorflow_estimator/python/estimator/canned/tensor_forest.py]: Yes, let's start out that way until we are ready to finalize the API-- estimator.experimental.TensorForestClassifier. Thanks--

Comment (yupbank) [reply tensorflow_estimator/python/estimator/canned/tensor_forest.py]: thank you thank you, let me do that

File: tensorflow_estimator/python/estimator/canned/tensor_forest.py (hunk 4 above)
Comment (yupbank) [reply tensorflow_estimator/python/estimator/canned/tensor_forest.py]: nice, so i don't need to check here right? but then how would the user pass in sparse_columns ?

**Comment by yupbank** [comment]
@tanzhenyu  is there a way for me to check the CI logs ?  right now i can't see anything but failed

image

File: tensorflow_estimator/python/estimator/experimental/tensor_forest.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+    if head is None:
+      # pylint: disable=protected-access
+      head = head_lib._binary_logistic_or_multi_class_head(
+          n_classes=n_classes,
+          weight_column=None,
+          label_vocabulary=label_vocabulary,
+          loss_reduction=losses.Reduction.SUM_OVER_BATCH_SIZE)
+    else:
+      assert head.logits_dimension == n_classes, 'head logits_dimension: %s ' \
+          % head.logits_dimension + \
```
Comment (nataliaponomareva) [inline tensorflow_estimator/python/estimator/experimental/tensor_forest.py]: i think this will fail for binary logistic head (logits dim 1, num classes =2). Maybe add if numclasses>2 for this check

File: tensorflow_estimator/python/estimator/experimental/tensor_forest.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+  estimator_spec.predictions[VARIANCE_LOGITS_KEY] = logits_variance
+  estimator_spec.predictions[
+      VARIANCE_AFTER_ACTIVATION_KEY] = var_after_activation
+
+  return estimator_spec
+
+
+# @estimator_export('estimator.TensorForestClassifier')
+class TensorForestClassifier(estimator.Estimator):
```
Comment (nataliaponomareva) [inline tensorflow_estimator/python/estimator/experimental/tensor_forest.py]: so there is a classifier but no regressor. I would suggest to either rename it to TensorForestEstimator, or adding a regressor as well (with l2 loss head)

Comment (nataliaponomareva) [reply tensorflow_estimator/python/estimator/experimental/tensor_forest.py]: In boosted, we have a classifier, a regressor and an estimator that takes an arbitrary head...

Comment (yupbank) [reply tensorflow_estimator/python/estimator/experimental/tensor_forest.py]: ah yeah, the regressor will be added. in a separate pr?

Comment (nataliaponomareva) [reply tensorflow_estimator/python/estimator/experimental/tensor_forest.py]: sg!

File: tensorflow_estimator/python/estimator/experimental/tensor_forest_test.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    self.assertTrue(isinstance(logits, ops.Tensor))
+    self.assertTrue(isinstance(var, ops.Tensor))
+    with self.test_session():
+      variables.global_variables_initializer().run()
+      resources.initialize_resources(
+          resources.shared_resources()).run()
+      self.assertEquals((4, 2), logits.eval().shape)
+      self.assertAllClose(expected_logits,
+                          logits.eval().tolist())
+
```
Comment (nataliaponomareva) [inline tensorflow_estimator/python/estimator/experimental/tensor_forest_test.py]: We can probably already add an estimator test. 
TestTensorForestClassifierDoesNotThrowException
est = TensorForestClassifier(...)
est.train(num_steps=1) # does not do anything for now
res = est.predict(input_fn..)
res contains logits, loss, var etc

**Comment by pragyaak** [comment]
@yupbank  gentle reminder to submit your response to the review

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
